feat: playwright browser tool

jisonZ opened the pull request:
# Description

Add playwright browser agent tool, most code are adapted from the corresponding tool in langchain: https://github.com/langchain-ai/langchain/tree/0c782ee5479f1eadf6cc532d7f27617bd2671fe9/libs/community/langchain_community/tools/playwright. 

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [x] Added example for testing functionality
- [x] I believe this change is already covered by existing unit tests

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

logan-markewich commented on llama-index-integrations/tools/llama-index-tools-playwright/llama_index/tools/playwright/base.py:
```diff
@@ -0,0 +1,355 @@
+from llama_index.core.tools.tool_spec.base import BaseToolSpec
+
+
+def lazy_import_playwright_browsers():
```
empty function?

logan-markewich commented on llama-index-integrations/tools/llama-index-tools-playwright/llama_index/tools/playwright/base.py:
```diff
@@ -0,0 +1,355 @@
+import json
+from urllib.parse import urlparse
+
+if TYPE_CHECKING:
```
Since playwright should be installed with this package, no need for the type checking imports here i think, we should be able to import directly.

jisonZ replied:
fixed in https://github.com/run-llama/llama_index/pull/17706/commits/32115c84cc8d105b0bec5898c52b0dd4ba0d7ef7

logan-markewich reviewed (approved):
This looks pretty good to me honestly, nice!

logan-markewich commented on llama-index-integrations/tools/llama-index-tools-playwright/tests/test_tools_playwright.py:
```diff
@@ -0,0 +1,7 @@
+
+def test_class():
+    names_of_base_classes = [b.__name__ for b in PlaywrightToolSpec.__mro__]
+    assert BaseToolSpec.__name__ in names_of_base_classes
```
Would be nice to have one or two actual unit tests 👁️

jisonZ commented:
@logan-markewich If the fixes look alright can we get the github actions going? It seems to stuck at "waiting for status to be reported", not sure if a maintainer permission is required here

logan-markewich commented:
@jisonZ yea first-time maintainers require manually trigger of the actions. A little annoying, but once this merges, future PRs for you will run automatically :)

jisonZ commented:
> @jisonZ yea first-time maintainers require manually trigger of the actions. A little annoying, but once this merges, future PRs for you will run automatically :)

The previous run was not successful since the unit tests require playwright browser install with `playwright install` additional to installing the playwright package. Forgot about that and added it to the poetry make file (hope this is the right place), would love to get approved for another workflow run

logan-markewich commented:
Ok, got tests to run, but tests fail? (also, seems like some other deps might be missing, lxml?)

jisonZ commented:
> Ok, got tests to run, but tests fail? (also, seems like some other deps might be missing, lxml?)

Added dependency to pyproject.toml, lets try again

logan-markewich commented:
Hmmm, 

`bs4.exceptions.FeatureNotFound: Couldn't find a tree builder with the features you requested: lxml. Do you need to install a parser library?`

Is lxml an extra on the bs4 package?

jisonZ commented:
> Hmmm,
> 
> `bs4.exceptions.FeatureNotFound: Couldn't find a tree builder with the features you requested: lxml. Do you need to install a parser library?`
> 
> Is lxml an extra on the bs4 package?

Yea it is suggested to have lxml parser library, it felt weird cuz i think i added `lxml = "^5.3.0"` to pyproject.toml and it should be working

jisonZ commented:
@logan-markewich according to this post: https://stackoverflow.com/questions/24398302/bs4-featurenotfound-couldnt-find-a-tree-builder-with-the-features-you-requeste, there are three parsers: `html.parser`, `lxml` and `html5lib`, and `html.parser` does not require any dependency. I've changed the default parser to `html.parser` and a parameter for user to set. It should no longer have dependency issues. The test cases are also updated.
